### DISCUSSION TITLE
TimeOfDay.when_activated doesn't seem to run

### DISCUSSION THREAD

**Discussion by chazapp** [gpiozero/gpiozero#1028] [Q&A]
Hello, i'm writing an automated garden monitoring application to which i'd like to add LED controls. I'm trying to use TimeOfDay to start/stop my LEDs automatically on a schedule. You can see my code [here](https://github.com/chazapp/garden/blob/master/lights.py).

The program is a Flask API exposing a `/metrics` endpoint to be scraped by Prometheus.
It seems that the `Lights.onActivate()` method I pass to TimeOfDay doesn't fire when the program starts and the LED do not light up. If I however run the method manually, the LEDs do light up, as per the example below:

```
# At that point, the application container is running, the metrics are showing up in Prometheus, but the LED lights are off
$ docker exec -it garden_garden_1 flask shell
/usr/local/lib/python3.9/site-packages/gpiozero/pins/pi.py:153: SPISoftwareFallback: failed to initialize hardware SPI, falling back to software (error was: failed to import spidev)
  warnings.warn(
Python 3.9.10 (main, Jan 29 2022, 05:27:31)
[GCC 10.2.1 20210110] on linux
App: app [production]
Instance: /app/instance
# The LED lights are off
>>> app.lights.red
<gpiozero.PWMLED object on pin GPIO16, active_high=True, is_active=False>
>>> app.lights.blue
<gpiozero.PWMLED object on pin GPIO21, active_high=True, is_active=False>
>>> app.lights.green
<gpiozero.PWMLED object on pin GPIO20, active_high=True, is_active=False>
>>> app.lights.onActivate()
# At that point, the LED do light up
>>> app.lights.daytime
<gpiozero.TimeOfDay object active between 08:00:00 and 20:00:00 UTC>
>>> app.lights.daytime.is_active
True
>>> app.lights.daytime.when_activated
<bound method Lights.onActivate of <lights.Lights object at 0x7f7dc586a0>>
>>> app.lights.daytime.when_deactivated()
>>>
>>> app.lights.red
<gpiozero.PWMLED object on pin GPIO16, active_high=True, is_active=False>
>>> app.lights.blue
<gpiozero.PWMLED object on pin GPIO21, active_high=True, is_active=False>
>>> app.lights.green
<gpiozero.PWMLED object on pin GPIO20, active_high=True, is_active=False>
>>> app.lights.daytime.is_active
True
>>> app.lights.daytime.when_activated()
>>> app.lights.red
<gpiozero.PWMLED object on pin GPIO16, active_high=True, is_active=True>
>>> app.lights.blue
<gpiozero.PWMLED object on pin GPIO21, active_high=True, is_active=True>
>>> app.lights.green
<gpiozero.PWMLED object on pin GPIO20, active_high=True, is_active=True>
```  
  
What could be the issue preventing my LEDs to be active on startup ?

**Comment by lurch** [comment]
Similar to #393, which was apparently fixed by #941 ? Perhaps you've found a corner-case which wasn't fully fixed?

Note also that GPIO Zero is designed to have "exclusive" control over the GPIO pins, i.e. running a GPIO Zero script in the background, and also interactively running GPIO Zero in a REPL at the same time, may lead to unexpected results.

EDIT: Also, you don't need to do both `self.red.value = 1` and `self.red.on()` - they're two different ways of doing the same thing.

**Reply by lurch** [reply]
Although _perhaps_ the 'problem' (I've not done any testing, so I may be completely wrong!) is that the [TimeOfDay](https://gpiozero.readthedocs.io/en/stable/api_internal.html#timeofday) only calls `onActivate` when the time ticks over from 07:59 to 08:00 and doesn't automatically call `onActivate` if you create the object between 08:00 and 20:00 ? If that's the case, I guess you could fix it by doing:
```python
        self.daytime = TimeOfDay(time(8), time(20))
        self.daytime.when_activated = self.onActivate
        self.daytime.when_deactivated = self.onDeactivate
        # set initial state as appropriate
        if self.daytime.is_active:
            self.daytime.when_activated()
        else:
            self.daytime.when_deactivated()
```

and you _might_ be able to reduce that to just
```python
        self.daytime = TimeOfDay(time(8), time(20))
        self.daytime.when_activated = self.onActivate
        self.daytime.when_deactivated = self.onDeactivate
        # set initial state as appropriate
        self.daytime._fire_events()
```
(see https://gpiozero.readthedocs.io/en/stable/api_generic.html#eventsmixin )
